### ISSUE TITLE
CustomLayer usage to add custom tiles as map layer

### ISSUE THREAD

**Issue by pahakorolev** [mapbox/mapbox-gl-native#4210]
MapBox Android SDK has methods to work with [`CustomLayers`](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/layers/CustomLayer.java):
```
com.mapbox.mapboxsdk.maps.MapboxMap.addCustomLayer(CustomLayer customLayer, String before)
com.mapbox.mapboxsdk.maps.MapboxMap.removeCustomLayer(String id)
com.mapbox.mapboxsdk.maps.MapboxMap.invalidateCustomLayers()
```
But using of those require native(cpp) implementation of CustomLayer like at your [example](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/android/src/example_custom_layer.cpp). 
It's so difficult, because we need to implement render and other parts of `CustomLayer`.
How about simple implementation for using raster tiles over rendered map like [`TileOverlay`](https://developers.google.com/android/reference/com/google/android/gms/maps/model/TileOverlay) and  [`UrlTileProvider`](https://developers.google.com/android/reference/com/google/android/gms/maps/model/UrlTileProvider) from Google Maps SDK for example?
It is necessary first of all to support the rendering tiles from server side, which can contains a lot of data like polygons, polylines or paths, markers and images already rendered by server.
Here examples of tiles:

![0539c9cef7e2e2b09c3145b03cdfc253 2](https://cloud.githubusercontent.com/assets/6101875/13553598/dd7f9e3a-e39f-11e5-8004-b6a180908233.png)

![0539c9cef7e2e2b09c3145b03cdfc253 1](https://cloud.githubusercontent.com/assets/6101875/13553601/ebbdab0e-e39f-11e5-964b-0b905092703c.png)

![0539c9cef7e2e2b09c3145b03cdfc253](https://cloud.githubusercontent.com/assets/6101875/13553604/f06c05ec-e39f-11e5-9efb-b574a842e0d7.png)

**Comment by 1ec5** [comment]
`CustomLayer` isn’t intended for this use case. It sounds like you want something like [this custom raster style example](https://www.mapbox.com/android-sdk/custom-raster-style/). If you want to overlay your raster tiles atop a GL style, such as Light, you can combine the existing [Light stylesheet](https://github.com/mapbox/mapbox-gl-styles/blob/v1.2.7/styles/light-v8.json) with the raster tile source in the provided example.

**Comment by pahakorolev** [comment]
No, that's not what I meant. The custom tile layer can be added dynamically, e.g. user pressed some button. In your case I need to change style url and reload all data, which previously was added. And user can edit this layer by some request to a server, e.g. change period of history which he is seeing on the map. And so, I should be able to invalidate this layer, because server's tiles was changed.
If need be, I'm ready to show necessity of this feature at real cases and applications.

**Comment by 1ec5** [comment]
Ah, thanks for the clarification. From the screenshots and explanation above, the raster tiles look simple enough that they could be represented more efficiently as GeoJSON rather than as raster images. (GeoJSON makes things like [mapbox-gl-draw](https://github.com/mapbox/mapbox-gl-draw) possible.) The Android SDK has support for GeoJSON at the source level. #3630 tracks dynamically modifying GeoJSON layers at runtime, while #837 tracks dynamically adding and removing layers, paint properties, etc. As far as I know, the combination of those two features would enable the use case you’re describing.

**Comment by pahakorolev** [comment]
Thank you for your answer. GeoJSON is a cool feature, but described by me functionality is not the same.
Firstly, which size of  GeoJson we'll have, if it will contains about 1 million points? But it may contain more.
And secondly, I speak about existing solution. Our server side returning png tiles and change it's to GeoJSON will forces us to change a significant amount of our logic.
And a possibility of working with custom tiles is a basic functionality of any popular cartographical SDK.
For example:
- [Google](https://developers.google.com/android/reference/com/google/android/gms/maps/GoogleMap.html#addTileOverlay(com.google.android.gms.maps.model.TileOverlayOptions)) Maps Android, and other platforms so on.
- [osmdroid](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/tileprovider/tilesource/XYTileSource.java)
- [ArcGis](https://developers.arcgis.com/android/api-reference/reference/com/esri/android/map/TiledServiceLayer.html)
- [Nutiteq](http://nutiteq.github.io/hellomap3d-android/com/nutiteq/layers/RasterTileLayer.html)

and etc.

**Comment by 1ec5** [comment]
In that case, I think #837 is the one you want to keep an eye on. Here’s what I mean: although [this GL JS example](https://www.mapbox.com/mapbox-gl-js/example/geojson-layer-in-stack/) uses a GeoJSON layer, there’s nothing that prevents you from calling `addLayer()` with a raster tile layer, and there are related methods like [`addSource()`](https://www.mapbox.com/mapbox-gl-js/api/#Map.addSource) and [`removeLayer()`](https://www.mapbox.com/mapbox-gl-js/api/#Map.removeLayer) as well.

**Comment by pahakorolev** [comment]
Thanks for the explanation. If by this style I will be able to use raster tiles from any custom server - it will be a good solution.
```
val layer=map.addLayer({
"sources": {
            "raster-tiles": {
                "type": "raster",
                "url": "http://some.server",
                "tileSize": 512
            }
        }
"tiles": [
        "http://some.server/{z}/{x}/{y}/tile.png"
    ]
});
...//some code
map.invalidateLayer(layer); //or map.removeLayer(layer) and then map.addLayer({})
map.removeLayer(layer);
```

**Comment by 1ec5** [comment]
Image source support (mapbox/mapbox-gl-native#1350) may also be relevant to your use case.

**Comment by 1ec5** [comment]
> In your case I need to change style url and reload all data, which previously was added.

I missed this detail – you want to explicitly _reload_ a raster tile source rather than adding and removing distinct raster tile sources at runtime. The iOS SDK does have a method for reloading the style, but it doesn’t bust the cache. (The only thing that currently busts the cache is the `Cache-Control` HTTP header: https://github.com/mapbox/mapbox-gl-native/issues/4446#issuecomment-200583990.) Before we can introduce a cache-busting mechanism, we need to add support to our own servers for refreshing; we’re tracking that work internally.

**Comment by cammace** [comment]
@pahakorolev we've now added the runtime style API and closed #837, this exposes new ways to add a raster or GeoJSON layer to the map. For examples, I'd recommend taking a look at [this ticket](https://github.com/mapbox/mapbox-android-demo/issues/115), there's a few examples showing how to add layers. To use in your application now, make sure you have the SNAPSHOT version of the SDK in your `build.gradle`.
